### ISSUE TITLE
multi class

### ISSUE THREAD

**Issue by taemin6697** [wgcban/ChangeFormer#42]
You want to perform multiclass classification.
Even if I change the code to args.n_class = 9, the class is predicted to be 2.
What should I do?
sry im korean
![캡처](https://user-images.githubusercontent.com/96530685/200052571-4145d5c8-d9cb-422b-9864-f6d453aadda3.PNG)
It shouldn't be possible to modify only n_class, but should there be multiple classes of labels?

**Comment by wgcban** [comment]
Please check this:
https://github.com/wgcban/ChangeFormer/issues/40#issuecomment-1295830619

**Comment by wgcban** [comment]
@taemin6697 You need to modify it here:
https://github.com/wgcban/ChangeFormer/blob/9025e26417cf8f10f29a48f34a05758498216465/models/networks.py#L159

Change this line to 
```
net = ChangeFormerV6(embed_dim=args.embed_dim, output_nc=args. n_class)
```

**Comment by taemin6697** [comment]
알겠습니다 한번 해보겠습니다 
Okay, I'll try

**Comment by taemin6697** [comment]
Should I train from scratch?

**Comment by wgcban** [comment]
@taemin6697 Yes, in this case, you can start training from scratch. Or else you can specify strict=false when loading the pre-trained model.

**Comment by taemin6697** [comment]
Even if I do strict=false I get the same error.

**Comment by taemin6697** [comment]
We are currently training like this.
print("Initializing backbone weights from: " + self.args.pretrain)
        self.net_G.load_state_dict(torch.load(self.args.pretrain), strict=False)
        self.net_G.to(self.device)
        self.net_G.eval()

I'm sorry, but now I don't have data to distinguish more than 3 classes, so I'm not sure if it trains when I run that.
  When n_class = 3 with LEVIR-CD data, hist gives an error.

I'm not sure if that works for datasets with multiple labels. sorry. thanks for the reply

![캡처](https://user-images.githubusercontent.com/96530685/200066937-0a250957-4ded-4d48-a8f4-48fde07716fb.PNG)

**Comment by wgcban** [comment]
@taemin6697 this error can be expected because when we change the number of classes, the shape of the final convolution layer changes. So you need to avoid those when initializing the model here: https://github.com/wgcban/ChangeFormer/blob/9025e26417cf8f10f29a48f34a05758498216465/models/trainer.py#L135-L138
You can try something like this:
```
current_model_dict =  self.net_G.state_dict()
loaded_state_dict = torch.load(os.path.join(self.checkpoint_dir, ckpt_name))
new_state_dict={k:v if v.size()==current_model_dict[k].size()  else  current_model_dict[k] for k,v in zip(current_model_dict.keys(), loaded_state_dict.values())}
self.net_G.load_state_dict(new_state_dict, strict=False)
```

**Comment by wgcban** [comment]
For more details on this issue, refer https://github.com/pytorch/pytorch/issues/40859

**Comment by taemin6697** [comment]
@wgcban self.net_G.load_state_dict(checkpoint['model_G_state_dict'])
Can I change it from here?

**Comment by wgcban** [comment]
@taemin6697 No I think you should replace all these:
```
checkpoint = torch.load(os.path.join(self.checkpoint_dir, ckpt_name), 
                         map_location=self.device) 
 # update net_G states 
 self.net_G.load_state_dict(checkpoint['model_G_state_dict']) 
```

**Comment by wgcban** [comment]
Yes, my bad. Delete L138 `(self.net_G.load_state_dict(checkpoint['model_G_state_dict'])` with the above.

**Comment by taemin6697** [comment]
![캡처](https://user-images.githubusercontent.com/96530685/200078314-d28d9e85-a493-4400-9e4c-dd6101810324.PNG)

**Comment by taemin6697** [comment]
@wgcban a okay!

**Comment by wgcban** [comment]
Because `checkpoint` uses later to initialize the optimizer, so we need it.

**Comment by taemin6697** [comment]
@wgcban This error occurs when n_class=4.
Sorry we don't currently have data for multi-classes, so maybe the dimensions are different. The current dataset is LEVIR Sorry.
It's currently 6am Korean time, so it's too late. I'll get up tomorrow and test it with a dataset for multiclass.

![캡처](https://user-images.githubusercontent.com/96530685/200078899-2364d0da-0fd7-4368-86ea-21ba7918a873.PNG)

**Comment by wgcban** [comment]
Sure.

**Comment by taemin6697** [comment]
@wgcban 
I tried it with a small data set just in case. I have 20 classes and it doesn't work well.
![캡처](https://user-images.githubusercontent.com/96530685/200085714-dd76a42f-ab87-42cd-9642-3e6340682d48.PNG)

**Comment by taemin6697** [comment]
@wgcban 
'Next, ensure that the ground truth png file contains class numbers (0,1,2,3) when read.'
I didn't quite understand what this meant. sorry.

**Comment by zdw9915** [comment]
Hello, sorry to bother you, after training the ChangeFormerV6 model with the levir_cd dataset, I am ready to use the trained ChangeFormerV6 model to make predictions, and then an error occurs, how to solve it?
![image](https://user-images.githubusercontent.com/113034050/205484112-08ec3f2a-ef6a-43de-ba88-28a73154b83a.png)

**Comment by wgcban** [comment]
@zdw9915 It seems like your model and the trained model has different parameters. Are you using the correct model during the evaluation?

**Comment by wangqingl** [comment]
> @wgcban I tried it with a small data set just in case. I have 20 classes and it doesn't work well. ![캡처](https://user-images.githubusercontent.com/96530685/200085714-dd76a42f-ab87-42cd-9642-3e6340682d48.PNG)

Hello, I have also encountered the same problem as you. Have you resolved it

**Comment by Hank-user** [comment]
> 您好，很抱歉打扰您，用 levir_cd 数据集训练了 ChangeFormerV6 模型后，准备使用训练好的 ChangeFormerV6 模型进行预测，然后出现错误，如何解决呢？ ![image](https://user-images.githubusercontent.com/113034050/205484112-08ec3f2a-ef6a-43de-ba88-28a73154b83a.png)
Hello! I have also encountered the same problem, may I ask how to solve this problem
